polkadot/src/main.rs

```diff
@@ -19,12 +19,47 @@
 #![warn(missing_docs)]
 
 extern crate polkadot_cli as cli;
+extern crate ctrlc;
+extern crate futures;
 
 #[macro_use]
 extern crate error_chain;
 
+use cli::{ClientError, ServiceComponents, ClientBackend, PolkadotBlock, StateMachineBackend, Service};
+use futures::sync::oneshot;
+use futures::{future, Future};
+
+use std::cell::RefCell;
+
+// the regular polkadot application simply runs until ctrl-c
+struct Application;
+impl cli::Application for Application {
+	type Work = future::Empty<(), ()>;
+	type Exit = future::MapErr<oneshot::Receiver<()>, fn(oneshot::Canceled) -> ()>;
+
+	fn exit(&mut self) -> Self::Exit {
+		// can't use signal directly here because CtrlC takes only `Fn`.
+		let (exit_send, exit) = oneshot::channel();
+
+		let exit_send_cell = RefCell::new(Some(exit_send));
+		ctrlc::CtrlC::set_handler(move || {
+			if let Some(exit_send) = exit_send_cell.try_borrow_mut().expect("signal handler not reentrant; qed").take() {
+				exit_send.send(()).expect("Error sending exit notification");
+			}
+		});
+
+		exit.map_err(drop)
+	}
+
+	fn work<C: ServiceComponents>(self, _service: &Service<C>) -> Self::Work
+		where ClientError: From<<<<C as ServiceComponents>::Backend as ClientBackend<PolkadotBlock>>::State as StateMachineBackend>::Error>,
+	{
+		future::empty()
+	}
+}
+
 quick_main!(run);
 
 fn run() -> cli::error::Result<()> {
-	cli::run(::std::env::args())
+	cli::run(::std::env::args(), Application)
```

Comment:
I think `cli::Application` is a little consuming name... Here is my experience

When I checked this file for the first time I had an impression that `Application` provides a main logic of, well, the application. Only until I found that `work` returns `future::empty()` for the `Application` I realized that my hypothesis was wrong.

Reply:
I think of it as "application on top of the node". Since the normal operation mode is to do no additional work _with the node_ then the work future is empty. But with the collator node the implementation of `Work` will be to listen to import notifications from the node and to push collated blocks to validators on the network.

I can't really think of a better name for it though.

Reply:
Yeah, figured it out pretty quick after that, so maybe not a big deal

I honestly was hoping that you will be able to come up with a better name : )

Reply:
(renaming to `Worker` on suggestion of @gnunicorn )